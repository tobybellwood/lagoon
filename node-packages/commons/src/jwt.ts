import { is } from 'ramda';
import { sign } from 'jsonwebtoken';

type Role = 'none' | 'admin' | 'drush';
interface Permissions {
  projects: number[];
  customers: number[];
}

interface Payload {
  role: Role;

  // Issuer - Information on who created this token
  iss: string;

  // Subject - Unique identifier for the user / identity
  sub?: string;

  // Audience - Url of the resource server the token was intended for (e.g. http://api.amazee.io)
  aud?: string;

  // IssuedAt timestamp (needs to be > 0)
  iat?: number;

  // Expiration as NumericDate
  // Example for an expiration timestamp in 1 hour: Math.floor(Date.now() / 1000) + (60 * 60)
  exp?: number;
}

interface PayloadWithUser extends Payload {
  userId: number;
  permissions: Permissions;
}

// Most of these parameters are described in more detail in the jsonwebtoken README
// See: https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback
interface CreateJWTCommonArgs {
  expiresIn?: string; //expressed in seconds or a string describing a time span zeit/ms. Eg: 60, "2 days", "10h", "7d"
  jwtSecret: string; // The passphrase to sign the token with
}

interface CreateJWTArgs extends CreateJWTCommonArgs {
  payload: PayloadWithUser;
}

interface CreateJWTWithoutUserIdArgs extends CreateJWTCommonArgs {
  payload: Payload;
}

// Create a token from given metadata and userId
export const createJWT = async ({
  payload,
  payload: { userId },
  jwtSecret,
  expiresIn
}: CreateJWTArgs): Promise<string> =>
  new Promise((resolve, reject) => {
    if (!is(Number, userId)) {
      reject(
        new Error(
          `Incorrect userId parameter "${userId}" passed (expecting a number)!`
        )
      );
    }

    // Sometimes we want some expiresIn values, if we don't know an exact payload.exp
    const options = expiresIn ? { expiresIn } : null;

    // By passing an object, sign will add an issued at timestamp (iat).
    // Because of this, the token hash will be different for every call of this function.
    sign(payload, jwtSecret, options, (err, token) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(token);
    });
  });

// Create a JWT without any user ID checks.
//
// This function is also synchronous, which is useful for creating service
// tokens, which generally get created during startup of services. Beware
// the performance implications of using this in async environments!
export const createJWTWithoutUserId = ({
  expiresIn,
  jwtSecret,
  payload
}: CreateJWTWithoutUserIdArgs): string => {
  // Sometimes we want some expiresIn values, if we don't know an exact payload.exp
  const options = expiresIn ? { expiresIn } : null;

  // By passing an object, sign will add an issued at timestamp (iat).
  // Because of this, the token hash will be different for every call of this function.
  return sign(payload, jwtSecret, options);
};
