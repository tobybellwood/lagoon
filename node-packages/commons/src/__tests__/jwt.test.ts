import jwt from 'jsonwebtoken';
import { createJWT } from '../jwt';

describe('createJWT', () => {
  test('should create token properly', async () => {
    const input = {
      payload: {
        userId: 0,
        permissions: {
          projects: [],
          customers: []
        },
        role: 'admin' as const,
        iat: 1503014400000,
        iss: 'jest',
      },
      jwtSecret: 'secret',
    };

    const token = await createJWT(input);
    const decodedPayload = jwt.verify(token, 'secret');

    expect(token).toMatchSnapshot();
    expect(decodedPayload).toEqual(input.payload);
  });

  test('should throw an error on invalid user id', () => {
    const input = {
      payload: {
        userId: 'invalid userId',
        permissions: {
          projects: [],
          customers: []
        },
        role: 'admin' as const,
        iat: 1503014400000,
        iss: 'jest',
      },
      jwtSecret: 'secret',
    };

    // @ts-ignore This intentionally passes an incorrect format to createJWT
    expect(createJWT(input)).rejects.toEqual(
      new Error(
        'Incorrect userId parameter "invalid userId" passed (expecting a number)!',
      ),
    );
  });

  test('should create token with expiration date via expiresIn argument', async () => {
    const hour = 3600;
    const currentTime = Math.floor(Date.now() / 1000);

    const input = {
      payload: {
        userId: 0,
        permissions: {
          projects: [],
          customers: []
        },
        role: 'admin' as const,
        iat: currentTime,
        iss: 'jest',
      },
      jwtSecret: 'secret',
      expiresIn: '1h',
    };

    const token = await createJWT(input);
    const p = jwt.verify(token, 'secret');

    expect(p.exp).toEqual(currentTime + hour);
  });

  test('should create expired token, which should be reckognized by verify()', async () => {
    const input = {
      payload: {
        userId: 0,
        permissions: {
          projects: [],
          customers: []
        },
        role: 'admin' as const,
        iat: 1503014400, // way back in the past
        iss: 'jest',
      },
      jwtSecret: 'secret',
      expiresIn: '1h',
    };

    const token = await createJWT(input);
    const fn = () => jwt.verify(token, 'secret');

    expect(fn).toThrow(new Error('jwt expired'));
  });
});
