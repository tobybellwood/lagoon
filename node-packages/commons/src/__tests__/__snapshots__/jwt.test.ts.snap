// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createJWT should create token properly 1`] = `"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjAsInBlcm1pc3Npb25zIjp7InByb2plY3RzIjpbXSwiY3VzdG9tZXJzIjpbXX0sInJvbGUiOiJhZG1pbiIsImlhdCI6MTUwMzAxNDQwMDAwMCwiaXNzIjoiamVzdCJ9.22DgaCLyG2IsbMmD9hkP1S_zWIh_XH8IJCIIHGpoLx0"`;
